--- contrib/systemd/openarc.service.in.orig	2026-03-13 20:47:24.896982050 +0900
+++ contrib/systemd/openarc.service.in	2026-03-13 20:47:24.900494231 +0900
@@ -10,11 +10,16 @@
 EnvironmentFile=-@sysconfdir@/sysconfig/openarc
 RuntimeDirectory=openarc
 RuntimeDirectoryMode=0750
-ExecStart=@sbindir@/openarc -f $OPTIONS
+StateDirectory=openarc
+StateDirectoryMode=0750
+ConfigurationDirectory=openarc
+ConfigurationDirectoryMode=0750
+ExecStart=@sbindir@/openarc -f -c /etc/openarc/openarc.conf
 ExecReload=/bin/kill -USR1 $MAINPID
 Restart=on-abnormal
 User=openarc
 Group=openarc
+UMask=0007
 ProtectSystem=strict
 ProtectHome=true
 
