diff --git a/nvidia-drm/nvidia-drm-gem-nvkms-memory.c b/nvidia-drm/nvidia-drm-gem-nvkms-memory.c
index e4114e0..32c31e8 100644
--- a/nvidia-drm/nvidia-drm-gem-nvkms-memory.c
+++ b/nvidia-drm/nvidia-drm-gem-nvkms-memory.c
@@ -333,9 +333,6 @@ int nv_drm_gem_import_nvkms_memory_ioctl(struct drm_device *dev,
 
     if (pMemory == NULL) {
         ret = -EINVAL;
-        NV_DRM_DEV_LOG_ERR(
-            nv_dev,
-            "Failed to import NVKMS memory to GEM object");
         goto nvkms_import_memory_failed;
     }
 
@@ -486,9 +483,6 @@ static struct drm_gem_object *__nv_drm_gem_nvkms_prime_dup(
     pMemory = nvKms->dupMemory(nv_dev->pDevice,
                                nv_dev_src->pDevice, nv_gem_src->pMemory);
     if (pMemory == NULL) {
-        NV_DRM_DEV_LOG_ERR(
-            nv_dev,
-            "Failed to import NVKMS memory to GEM object");
         goto nvkms_dup_memory_failed;
     }
 
